### PR TITLE
update Bernstein pdf

### PR DESCRIPTION
**Pull request by rahulgrit**

### CODE REVIEW

File: roofit/roofit/inc/RooBernstein.h [hunk 1]
```diff
@@ -33,11 +33,13 @@ class RooBernstein : public RooAbsPdf {
 
   Int_t getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVars, const char* rangeName=0) const ;
   Double_t analyticalIntegral(Int_t code, const char* rangeName=0) const ;
+  virtual void selectNormalizationRange(const char* rangeName=0, Bool_t force=kFALSE) ;
```
Comment (hageboeck) [inline roofit/roofit/inc/RooBernstein.h:36]:
```suggestion
void selectNormalizationRange(const char* rangeName=0, Bool_t force=kFALSE) ;
```
Let's make it virtual only when needed, i.e., when somebody decides to derive from RooBernstein.

File: roofit/roofit/inc/RooBernstein.h [hunk 2]
```diff
@@ -33,11 +33,13 @@ class RooBernstein : public RooAbsPdf {
...
 
 private:
 
   RooRealProxy _x;
   RooListProxy _coefList ;
+  const TNamed* _refRangeName ;
```
Comment (hageboeck) [inline roofit/roofit/inc/RooBernstein.h:42]: No need to put a pointer, just use simple string.
```suggestion
 std::string _refRangeName ;
```

File: roofit/roofit/src/RooBernstein.cxx [hunk 3]
```diff
@@ -47,7 +47,7 @@ ClassImp(RooBernstein);
 
 ////////////////////////////////////////////////////////////////////////////////
 
-RooBernstein::RooBernstein()
+RooBernstein::RooBernstein() : _refRangeName(nullptr)
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:50]:
```suggestion
RooBernstein::RooBernstein()
```
With simple string, don't need to to anything here.

File: roofit/roofit/src/RooBernstein.cxx [hunk 4]
```diff
@@ -58,7 +58,8 @@ RooBernstein::RooBernstein(const char* name, const char* title,
                            RooAbsReal& x, const RooArgList& coefList):
   RooAbsPdf(name, title),
   _x("x", "Dependent", this, x),
-  _coefList("coefficients","List of coefficients",this)
+  _coefList("coefficients","List of coefficients",this),
+  _refRangeName(nullptr)
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:62]:
```suggestion
  _coefList("coefficients","List of coefficients",this)
```

File: roofit/roofit/src/RooBernstein.cxx [hunk 5]
```diff
@@ -78,16 +79,32 @@ RooBernstein::RooBernstein(const char* name, const char* title,
 RooBernstein::RooBernstein(const RooBernstein& other, const char* name) :
   RooAbsPdf(other, name),
   _x("x", this, other._x),
-  _coefList("coefList",this,other._coefList)
+  _coefList("coefList",this,other._coefList),
+  _refRangeName(other._refRangeName)
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:83]:
```suggestion
  _coefList("coefList",this,other._coefList)
```

File: roofit/roofit/src/RooBernstein.cxx [hunk 6]
```diff
@@ -78,16 +79,32 @@ RooBernstein::RooBernstein(const char* name, const char* title,
...
 
+void RooBernstein::selectNormalizationRange(const char* rangeName, Bool_t force)
+{
+  if (rangeName && (force || !_refRangeName)) {
+     _refRangeName = static_cast<const TNamed*> (RooNameReg::instance().constPtr(rangeName));
+  }
+  if (!rangeName) {
+     _refRangeName = nullptr;
+  }
+}
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:97]: Use `_refRangeName.empty()`.

File: roofit/roofit/src/RooBernstein.cxx [hunk 7]
```diff
@@ -78,16 +79,32 @@ RooBernstein::RooBernstein(const char* name, const char* title,
...
+
+////////////////////////////////////////////////////////////////////////////////
+
 Double_t RooBernstein::evaluate() const
 {
-  Double_t xmin = _x.min();
-  Double_t x = (_x - xmin) / (_x.max() - xmin); // rescale to [0,1]
+  const Double_t xmax = _x.max(_refRangeName?_refRangeName->GetName():0);
+  const Double_t xmin = _x.min(_refRangeName?_refRangeName->GetName():0);
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:104]: Also here `.empty`.

File: roofit/roofit/src/RooBernstein.cxx [hunk 8]
```diff
@@ -78,16 +79,32 @@ RooBernstein::RooBernstein(const char* name, const char* title,
...
+  Double_t x = (_x - xmin) / (xmax - xmin); // rescale to [0,1]
+//  std::cout << "inside evaluate" << "xmin=" << xmin << ",xmax=" << xmax << std::endl;
+//  std::cout << "inside evaluate" << "_x.xmin(0)" << _x.min(0) << "_xmax(0)" << _x.max(0) << std::endl;
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:107]: Remove.

File: roofit/roofit/src/RooBernstein.cxx [hunk 9]
```diff
@@ -234,7 +255,8 @@ Double_t RooBernstein::analyticalIntegral(Int_t code, const char* rangeName) con
     // where the integral is straight forward.
     temp = 0;
     for (int j=i; j<=degree; ++j){ // power basis≈ß
-      temp += pow(-1.,j-i) * TMath::Binomial(degree, j) * TMath::Binomial(j,i) / (j+1);
+      temp += pow(-1.,j-i) * TMath::Binomial(degree, j) * TMath::Binomial(j,i) * (TMath::Power(xhi,j+1) - TMath::Power(xlo,j+1)) / (j+1);
+ //   temp += pow(-1.,j-i) * TMath::Binomial(degree, j) * TMath::Binomial(j,i) / (j+1);
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:259]: Remove.

File: roofit/roofit/src/RooBernstein.cxx [hunk 10]
```diff
@@ -84,10 +84,20 @@ RooBernstein::RooBernstein(const RooBernstein& other, const char* name) :
 
 ////////////////////////////////////////////////////////////////////////////////
 
+void RooBernstein::selectNormalizationRange(const char* rangeName, Bool_t force)
+{
+  if (rangeName && (force || !_refRangeName.empty())) {
+     _refRangeName = static_cast<const TNamed*> (RooNameReg::instance().constPtr(rangeName));
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:90]:
```suggestion
     _refRangeName = rangeName;
```

File: roofit/roofit/src/RooBernstein.cxx [hunk 11]
```diff
@@ -84,10 +84,20 @@ RooBernstein::RooBernstein(const RooBernstein& other, const char* name) :
...
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
 Double_t RooBernstein::evaluate() const
 {
-  Double_t xmin = _x.min();
-  Double_t x = (_x - xmin) / (_x.max() - xmin); // rescale to [0,1]
+  const Double_t xmax = _x.max(_refRangeName?_refRangeName.c_str():0);
+  const Double_t xmin = _x.min(_refRangeName?_refRangeName.c_str():0);
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:99]: With new RooFit, you can do this in a single function call.
```suggestion
  double xmin, xmax;
  std::tie(xmin, xmax) = _x.getRange(_refRangeName.empty() ? nullptr : _refRangeName.c_str());
```
Same in the integral function.

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-performance-centos7-multicore/default

File: roofit/roofit/src/RooBernstein.cxx [hunk 12]
```diff
@@ -222,7 +232,11 @@ Int_t RooBernstein::getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVar
 Double_t RooBernstein::analyticalIntegral(Int_t code, const char* rangeName) const
 {
   R__ASSERT(code==1) ;
-  Double_t xmin = _x.min(rangeName); Double_t xmax = _x.max(rangeName);
+  const Double_t xmax = _x.max(_refRangeName.empty()?_refRangeName.c_str():0);
+  const Double_t xmin = _x.min(_refRangeName.empty()?_refRangeName.c_str():0);
```
Comment (hageboeck) [inline roofit/roofit/src/RooBernstein.cxx:236]: Would suggest to do the same with `std::tie` here.

File: roofit/roofit/src/RooBernstein.cxx (hunk 11 above)
Comment (rahulgrit) [reply roofit/roofit/src/RooBernstein.cxx]: Hi Stephan,

    When I test this change, I get the following error whilst compiling root.

error: 'const RooRealProxy {aka const class RooProxy<RooAbsReal>}' has no member named 'getRange'; did you mean 'hasRange'?
   std::tie(xmin, xmax) = _x.getRange(_refRangeName.empty() ? nullptr : _refRangeName.c_str());

**Comment by hageboeck** [comment]
@phsft-bot build

**Pull request closed without merging**